hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -422,9 +422,23 @@ private void submitConnectToClusterTask() {
                             logger.warning("No connection to cluster: " + clusterId);
                         }
 
-                        submitConnectToClusterTask();
+                        submitConnectToClusterTask(initialConnection);
                     }
                 }
+                if (initialConnection && !activeConnections.isEmpty()) {
+                    // Send state to cluster for async clients in async way.
+                    executor.submit(() -> {
+                        try {
+                            client.sendStateToCluster();
```

Comment:
I think there is still a problem here. Imagine we connected to the cluster with the `doConnectToCluster` call above. We see that `initialConnection && !activeConnections.isEmpty()` check is true, so we enter the if block, and let's assume we have successfully sent the state. Right here, imagine a big pause (just before the line `clientState = ClientState.INITIALIZED_ON_CLUSTER;`). And, at the same time, let's assume we disconnect from the cluster, and reconnect, so we are trying to send the state again (not here, but via `initializeClientOnCluster` method). Now, if this line is executed before the state is sent, we will be in trouble because we will signal that the client is in the `INITIALIZED_ON_CLUSTER` state, and ready for invocations, but it is in fact not.

Reply:
I assume synchronized(clientStateMutex) will be enough.